// angular-responsive-carousel/bundles/angular-responsive-carousel.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-responsive-carousel', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory(global['angular-responsive-carousel'] = {}, global.ng.core, global.ng.common));
}(this, (function (exports, core, common) { 'use strict';

    var Touches = /** @class */ (function () {
        function Touches(properties) {
            var _this = this;
            this.eventType = undefined;
            this.handlers = {};
            this.startX = 0;
            this.startY = 0;
            this.lastTap = 0;
            this.doubleTapMinTimeout = 300;
            this.tapMinTimeout = 200;
            this.touchstartTime = 0;
            this.i = 0;
            this.isMousedown = false;
            this._touchListeners = {
                "touchstart": "handleTouchstart",
                "touchmove": "handleTouchmove",
                "touchend": "handleTouchend"
            };
            this._mouseListeners = {
                "mousedown": "handleMousedown",
                "mousemove": "handleMousemove",
                "mouseup": "handleMouseup",
                "wheel": "handleWheel"
            };
            this._otherListeners = {
                "resize": "handleResize"
            };
            /*
             * Listeners
             */
            /* Touchstart */
            this.handleTouchstart = function (event) {
                _this.elementPosition = _this.getElementPosition();
                _this.touchstartTime = new Date().getTime();
                if (_this.eventType === undefined) {
                    _this.getTouchstartPosition(event);
                }
                _this.runHandler("touchstart", event);
            };
            /* Touchmove */
            this.handleTouchmove = function (event) {
                var touches = event.touches;
                // Pan
                if (_this.detectPan(touches)) {
                    _this.runHandler("pan", event);
                }
                // Pinch
                if (_this.detectPinch(event)) {
                    _this.runHandler("pinch", event);
                }
                // Linear swipe
                switch (_this.detectLinearSwipe(event)) {
                    case "horizontal-swipe":
                        event.swipeType = "horizontal-swipe";
                        _this.runHandler("horizontal-swipe", event);
                        break;
                    case "vertical-swipe":
                        event.swipeType = "vertical-swipe";
                        _this.runHandler("vertical-swipe", event);
                        break;
                }
                // Linear swipe
                if (_this.detectLinearSwipe(event) ||
                    _this.eventType === 'horizontal-swipe' ||
                    _this.eventType === 'vertical-swipe') {
                    _this.handleLinearSwipe(event);
                }
            };
            /* Touchend */
            this.handleTouchend = function (event) {
                var touches = event.touches;
                // Double Tap
                if (_this.detectDoubleTap()) {
                    _this.runHandler("double-tap", event);
                }
                // Tap
                _this.detectTap();
                _this.runHandler("touchend", event);
                _this.eventType = 'touchend';
                if (touches && touches.length === 0) {
                    _this.eventType = undefined;
                    _this.i = 0;
                }
            };
            /* Mousedown */
            this.handleMousedown = function (event) {
                _this.isMousedown = true;
                _this.elementPosition = _this.getElementPosition();
                _this.touchstartTime = new Date().getTime();
                if (_this.eventType === undefined) {
                    _this.getMousedownPosition(event);
                }
                _this.runHandler("mousedown", event);
            };
            /* Mousemove */
            this.handleMousemove = function (event) {
                //event.preventDefault();
                if (!_this.isMousedown) {
                    return;
                }
                // Pan
                _this.runHandler("pan", event);
                // Linear swipe
                switch (_this.detectLinearSwipe(event)) {
                    case "horizontal-swipe":
                        event.swipeType = "horizontal-swipe";
                        _this.runHandler("horizontal-swipe", event);
                        break;
                    case "vertical-swipe":
                        event.swipeType = "vertical-swipe";
                        _this.runHandler("vertical-swipe", event);
                        break;
                }
                // Linear swipe
                if (_this.detectLinearSwipe(event) ||
                    _this.eventType === 'horizontal-swipe' ||
                    _this.eventType === 'vertical-swipe') {
                    _this.handleLinearSwipe(event);
                }
            };
            /* Mouseup */
            this.handleMouseup = function (event) {
                // Tap
                _this.detectTap();
                _this.isMousedown = false;
                _this.runHandler("mouseup", event);
                _this.eventType = undefined;
                _this.i = 0;
            };
            /* Wheel */
            this.handleWheel = function (event) {
                _this.runHandler("wheel", event);
            };
            /* Resize */
            this.handleResize = function (event) {
                _this.runHandler("resize", event);
            };
            this.properties = properties;
            this.element = this.properties.element;
            this.elementPosition = this.getElementPosition();
            this.toggleEventListeners('addEventListener');
        }
        Object.defineProperty(Touches.prototype, "touchListeners", {
            get: function () {
                return this.properties.touchListeners ? this.properties.touchListeners : this._touchListeners;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Touches.prototype, "mouseListeners", {
            get: function () {
                return this.properties.mouseListeners ? this.properties.mouseListeners : this._mouseListeners;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Touches.prototype, "otherListeners", {
            get: function () {
                return this.properties.otherListeners ? this.properties.otherListeners : this._otherListeners;
            },
            enumerable: false,
            configurable: true
        });
        Touches.prototype.destroy = function () {
            this.toggleEventListeners('removeEventListener');
        };
        Touches.prototype.toggleEventListeners = function (action) {
            var listeners;
            if (this.properties.listeners === 'mouse and touch') {
                listeners = Object.assign(this.touchListeners, this.mouseListeners);
            }
            else {
                listeners = this.detectTouchScreen() ? this.touchListeners : this.mouseListeners;
            }
            if (this.properties.resize) {
                listeners = Object.assign(listeners, this.otherListeners);
            }
            for (var listener in listeners) {
                var handler = listeners[listener];
                // Window
                if (listener === "resize") {
                    if (action === 'addEventListener') {
                        window.addEventListener(listener, this[handler], false);
                    }
                    if (action === 'removeEventListener') {
                        window.removeEventListener(listener, this[handler], false);
                    }
                    // Document
                }
                else if (listener === 'mouseup' || listener === "mousemove") {
                    if (action === 'addEventListener') {
                        document.addEventListener(listener, this[handler], false);
                    }
                    if (action === 'removeEventListener') {
                        document.removeEventListener(listener, this[handler], false);
                    }
                    // Element
                }
                else {
                    if (action === 'addEventListener') {
                        this.element.addEventListener(listener, this[handler], false);
                    }
                    if (action === 'removeEventListener') {
                        this.element.removeEventListener(listener, this[handler], false);
                    }
                }
            }
        };
        Touches.prototype.addEventListeners = function (listener, handler) {
            window.addEventListener(listener, this[handler], false);
        };
        Touches.prototype.removeEventListeners = function (listener, handler) {
            window.removeEventListener(listener, this[handler], false);
        };
        Touches.prototype.handleLinearSwipe = function (event) {
            //event.preventDefault();
            this.i++;
            if (this.i > 3) {
                this.eventType = this.getLinearSwipeType(event);
            }
            if (this.eventType === 'horizontal-swipe') {
                this.runHandler('horizontal-swipe', event);
            }
            if (this.eventType === 'vertical-swipe') {
                this.runHandler('vertical-swipe', event);
            }
        };
        Touches.prototype.runHandler = function (eventName, response) {
            if (this.handlers[eventName]) {
                this.handlers[eventName](response);
            }
        };
        /*
         * Detection
         */
        Touches.prototype.detectPan = function (touches) {
            return touches.length === 1 && !this.eventType || this.eventType === 'pan';
        };
        Touches.prototype.detectDoubleTap = function () {
            var _this = this;
            if (this.eventType != undefined) {
                return;
            }
            var currentTime = new Date().getTime();
            var tapLength = currentTime - this.lastTap;
            clearTimeout(this.doubleTapTimeout);
            if (tapLength < this.doubleTapMinTimeout && tapLength > 0) {
                return true;
            }
            else {
                this.doubleTapTimeout = setTimeout(function () {
                    clearTimeout(_this.doubleTapTimeout);
                }, this.doubleTapMinTimeout);
            }
            this.lastTap = currentTime;
        };
        Touches.prototype.detectTap = function () {
            if (this.eventType != undefined) {
                return;
            }
            var currentTime = new Date().getTime();
            var tapLength = currentTime - this.touchstartTime;
            if (tapLength > 0) {
                if (tapLength < this.tapMinTimeout) {
                    this.runHandler("tap", event);
                }
                else {
                    this.runHandler("longtap", event);
                }
            }
        };
        Touches.prototype.detectPinch = function (event) {
            var touches = event.touches;
            return (touches.length === 2 && this.eventType === undefined) || this.eventType === 'pinch';
        };
        Touches.prototype.detectLinearSwipe = function (event) {
            var touches = event.touches;
            if (touches) {
                if (touches.length === 1 && !this.eventType || this.eventType === 'horizontal-swipe' || this.eventType === 'vertical-swipe') {
                    return this.getLinearSwipeType(event);
                }
            }
            else {
                if (!this.eventType || this.eventType === 'horizontal-swipe' || this.eventType === 'vertical-swipe') {
                    return this.getLinearSwipeType(event);
                }
            }
        };
        Touches.prototype.getLinearSwipeType = function (event) {
            if (this.eventType !== 'horizontal-swipe' && this.eventType !== 'vertical-swipe') {
                var movementX = Math.abs(this.moveLeft(0, event) - this.startX);
                var movementY = Math.abs(this.moveTop(0, event) - this.startY);
                if ((movementY * 3) > movementX) {
                    return 'vertical-swipe';
                }
                else {
                    return 'horizontal-swipe';
                }
            }
            else {
                return this.eventType;
            }
        };
        Touches.prototype.getElementPosition = function () {
            return this.element.getBoundingClientRect();
        };
        Touches.prototype.getTouchstartPosition = function (event) {
            this.startX = event.touches[0].clientX - this.elementPosition.left;
            this.startY = event.touches[0].clientY - this.elementPosition.top;
        };
        Touches.prototype.getMousedownPosition = function (event) {
            this.startX = event.clientX - this.elementPosition.left;
            this.startY = event.clientY - this.elementPosition.top;
        };
        Touches.prototype.moveLeft = function (index, event) {
            var touches = event.touches;
            if (touches) {
                return touches[index].clientX - this.elementPosition.left;
            }
            else {
                return event.clientX - this.elementPosition.left;
            }
        };
        Touches.prototype.moveTop = function (index, event) {
            var touches = event.touches;
            if (touches) {
                return touches[index].clientY - this.elementPosition.top;
            }
            else {
                return event.clientY - this.elementPosition.top;
            }
        };
        Touches.prototype.detectTouchScreen = function () {
            var prefixes = ' -webkit- -moz- -o- -ms- '.split(' ');
            var mq = function (query) {
                return window.matchMedia(query).matches;
            };
            if (('ontouchstart' in window)) {
                return true;
            }
            // include the 'heartz' as a way to have a non matching MQ to help terminate the join
            // https://git.io/vznFH
            var query = ['(', prefixes.join('touch-enabled),('), 'heartz', ')'].join('');
            return mq(query);
        };
        /* Public properties and methods */
        Touches.prototype.on = function (event, handler) {
            if (event) {
                this.handlers[event] = handler;
            }
        };
        return Touches;
    }());

    var Carousel = /** @class */ (function () {
        function Carousel(properties) {
            var _this = this;
            this.initialPositionX = 0;
            /* The index of the new position relative to the active index, for example -1 or +1 */
            this.newContainerPositionIndex = 0;
            this.preliminarySlideCounter = 0;
            /* Scrolling Index (counter) */
            this.slideCounter = 0;
            this.previousSlideCounter = 0;
            this.isContentImages = true;
            this.isLazyLoad = true;
            this.isContainerLocked = true;
            this.alignCells = "left";
            this.initialContainerPosition = 0;
            this.handleTouchstart = function (event) {
                _this.isTouchstart = true;
                _this.startX = _this.getStartX(event);
                _this.containerInitialPositionX = _this.getElementPosition()['left'] - _this.getCarouselElementPosition()['left'];
                _this.isMoveInProgress = true;
                _this.stopTransformContainer();
            };
            this.handleHorizontalSwipe = function (event) {
                var touches = event.touches;
                if (touches) {
                    _this.moveX = touches[0].clientX - _this.getCarouselElementPosition()['left'];
                }
                else {
                    _this.moveX = event.clientX - _this.getCarouselElementPosition()['left'];
                }
                _this.distanceAbs = _this.getDistanceAbs();
                _this.direction = _this.getDirection();
                _this.moveContainer();
            };
            this.handleTouchend = function (event) {
                if (!_this.isTouchstart) {
                    _this.isTouchstart = false;
                    return;
                }
                _this.isMoveInProgress = false;
                if (_this.detectSlide()) {
                    _this.handleSlide();
                }
                else {
                    _this.newContainerPositionIndex = 0;
                    _this.alignContainer();
                }
                _this.startX = _this.moveX = _this.distanceAbs = undefined;
                _this.isTouchstart = false;
            };
            this.element = properties.element;
            this.properties = properties;
            this.cells = this.element.children;
            this.visibleWidth = this.properties.visibleWidth || this.element.parentElement.clientWidth;
            this.setContainerWidth();
            this.alignContainer(0);
        }
        Object.defineProperty(Carousel.prototype, "cellLength", {
            get: function () {
                if (this.images) {
                    return this.images.length;
                }
                else {
                    return this.cells.length;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "totalContainerCellsCount", {
            get: function () {
                if (this.images) {
                    var cellLength = this.visibleCellsCount + this.overflowCellsLimit * 2;
                    if (cellLength > this.images.length) {
                        cellLength = this.images.length;
                    }
                    return cellLength;
                }
                else {
                    return this.cellLength;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "isFirstCell", {
            get: function () {
                return this.slideCounter === 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "lastCellIndex", {
            get: function () {
                return this.images.length ? (this.images.length - 1) : (this.cells.length - 1);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "overflowCellsLimit", {
            get: function () {
                if (this.images && this.isImagesLessCellLimit) {
                    return Math.floor((this.images.length - this.visibleCellsCount) / 2);
                }
                else {
                    return this.properties.overflowCellsLimit;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "isImagesLessCellLimit", {
            get: function () {
                return this.properties.overflowCellsLimit * 2 + this.visibleCellsCount > this.images.length;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "cellLimit", {
            get: function () {
                return this.visibleCellsCount + this.overflowCellsLimit * 2;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "images", {
            get: function () {
                return this.properties.images;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "margin", {
            get: function () {
                return this.properties.margin;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "minSwipeDistance", {
            get: function () {
                return this.properties.minSwipeDistance;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "transitionDuration", {
            get: function () {
                return this.properties.transitionDuration;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "transitionTimingFunction", {
            get: function () {
                return this.properties.transitionTimingFunction;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "fullCellWidth", {
            get: function () {
                return this.properties.cellWidth + this.margin;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "visibleCellsCount", {
            get: function () {
                return Math.ceil(this.visibleWidth / this.fullCellWidth);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "lapCounter", {
            get: function () {
                return Math.floor(this.slideCounter / this.totalContainerCellsCount);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(Carousel.prototype, "visibleCellsOverflowContainer", {
            get: function () {
                return (this.visibleCellsCount * this.fullCellWidth - this.margin) > this.visibleWidth;
            },
            enumerable: false,
            configurable: true
        });
        Carousel.prototype.stopTransformContainer = function () {
            this.transformPositionX(this.containerInitialPositionX, 0);
            this.setSlideCounter();
            this.isSlideInProgress = false;
            this.newContainerPositionIndex = 0;
            this.isSlideLengthLimited = undefined;
        };
        Carousel.prototype.getStartX = function (event) {
            var touches = event.touches;
            var startX;
            if (touches) {
                startX = touches[0].clientX - this.getCarouselElementPosition()['left'];
            }
            else {
                startX = event.clientX - this.getCarouselElementPosition()['left'];
            }
            return startX;
        };
        /* Move */
        Carousel.prototype.moveContainer = function () {
            var positionX = this.getMovePositionX();
            this.transformPositionX(positionX, 0);
        };
        Carousel.prototype.getMovePositionX = function () {
            var distance = this.getDistance();
            return this.containerInitialPositionX - distance;
        };
        /* Align */
        Carousel.prototype.alignContainer = function (duration) {
            if (duration === void 0) { duration = this.transitionDuration; }
            var positionX = this.getContainerPosition();
            this.transformPositionX(positionX, duration);
            this.setInitialContainerPosition(positionX);
        };
        Carousel.prototype.getContainerPosition = function () {
            var correction = this.getContainerPositionCorrection();
            this.isContainerPositionCorrection = correction != 0;
            return (this.initialContainerPosition + this.newContainerPositionIndex * this.fullCellWidth) + correction;
        };
        Carousel.prototype.setInitialContainerPosition = function (position) {
            var correction = this.getContainerPositionCorrection();
            this.initialContainerPosition = position - correction;
        };
        /* Offset the container to show the last cell completely */
        Carousel.prototype.getContainerPositionCorrection = function () {
            var correction = 0;
            if (this.properties.loop) {
                return 0;
            }
            if ((this.cellLength - this.preliminarySlideCounter) < this.visibleCellsCount || this.isSlideLengthLimited) {
                if (this.visibleWidth < this.totalContainerCellsCount * this.fullCellWidth) {
                    correction = -(this.visibleCellsCount * this.fullCellWidth - this.visibleWidth - this.margin);
                }
                if (correction >= -this.margin) {
                    correction = 0;
                }
            }
            return correction;
        };
        /* Quickly center */
        Carousel.prototype.quicklyPositionContainer = function () {
            var correction = this.getContainerPositionCorrection();
            var initialPosition = this.getNewContainerPosition() + correction;
            this.transformPositionX(initialPosition, 0);
            this.setInitialContainerPosition(initialPosition);
        };
        Carousel.prototype.getNewContainerPosition = function () {
            if (this.slideCounter > this.overflowCellsLimit) {
                if (this.alignCells === "left") {
                    this.lineUpCells();
                    return -(this.overflowCellsLimit * this.fullCellWidth);
                }
            }
            if (this.slideCounter <= this.overflowCellsLimit) {
                if (this.previousSlideCounter > this.overflowCellsLimit) {
                    this.lineUpCells();
                }
                return -(this.slideCounter * this.fullCellWidth);
            }
        };
        Carousel.prototype.setContainerWidth = function () {
            var containerWidth = this.getContainerWidth();
            this.element.style.width = containerWidth + "px";
        };
        Carousel.prototype.getContainerWidth = function () {
            var containerWidth = this.totalContainerCellsCount * this.fullCellWidth;
            var totalImageWidth = this.cellLength * this.fullCellWidth;
            if (totalImageWidth < containerWidth) {
                containerWidth = totalImageWidth;
            }
            return containerWidth;
        };
        Carousel.prototype.getFile = function (cellIndex) {
            var imageIndex = this.getFileIndex(cellIndex);
            var file = this.images[imageIndex];
            if (file && !file.type) {
                file.type = 'image';
            }
            return {
                image: this.images[imageIndex],
                imageIndex: imageIndex
            };
        };
        Carousel.prototype.getFileIndex = function (cellIndex) {
            var positionIndex = this.getPositionIndex(cellIndex);
            var numberLeftCells = (this.totalContainerCellsCount - 1) / 2;
            var imageIndex;
            var counter = this.slideCounter;
            if (!this.isContainerLocked) {
                if (counter <= this.overflowCellsLimit) {
                    return cellIndex;
                }
                else {
                    var cellLimitOverflow = counter - this.overflowCellsLimit;
                    imageIndex = positionIndex + cellLimitOverflow;
                    if (this.images && this.properties.loop) {
                        imageIndex = imageIndex % this.images.length;
                    }
                    return imageIndex;
                }
            }
            if (this.alignCells === "left") {
                if (counter > this.overflowCellsLimit) {
                    var cellLimitOverflow = counter - this.overflowCellsLimit;
                    imageIndex = positionIndex + cellLimitOverflow;
                }
                else {
                    imageIndex = cellIndex;
                }
            }
            if (imageIndex > this.lastCellIndex && !this.properties.loop) {
                return false;
            }
            return imageIndex;
        };
        Carousel.prototype.transformPositionX = function (value, duration) {
            if (duration === void 0) { duration = this.transitionDuration; }
            this.element.style.transition = 'transform ' + duration + 'ms ' + this.transitionTimingFunction;
            this.element.style.transform = 'translateX(' + value + 'px)';
        };
        Carousel.prototype.detectSlide = function () {
            return this.distanceAbs >= this.minSwipeDistance;
        };
        Carousel.prototype.next = function (length) {
            if (length === void 0) { length = 1; }
            this.newContainerPositionIndex = 0;
            this.direction = 'left';
            this.handleSlide(length);
        };
        Carousel.prototype.prev = function (length) {
            if (length === void 0) { length = 1; }
            this.newContainerPositionIndex = 0;
            this.direction = 'right';
            this.handleSlide(length);
        };
        Carousel.prototype.select = function (index) {
            this.slideCounter = index;
            this.quicklyPositionContainer();
        };
        Carousel.prototype.handleSlide = function (customSlideLength) {
            if (customSlideLength === void 0) { customSlideLength = undefined; }
            var isUsingButton = customSlideLength;
            if (isUsingButton && this.isSlideInProgress) {
                return;
            }
            this.slideLength = customSlideLength ? customSlideLength : this.limitSlideLength(this.getSlideLength());
            if (this.direction === 'left') {
                this.handleLeftSlide();
            }
            if (this.direction === 'right') {
                this.handleRightSlide();
            }
            this.alignContainer();
        };
        Carousel.prototype.handleLeftSlide = function () {
            this.preliminarySlideCounter = this.slideCounter + this.slideLength;
            var isLastSlide = this.detectLastSlide(this.slideCounter + this.slideLength);
            if (!isLastSlide) {
                this.newContainerPositionIndex = this.newContainerPositionIndex - this.slideLength;
                this.isSlideInProgress = true;
                if (this.isLazyLoad) {
                    this.isContainerLocked = this.preliminarySlideCounter > this.overflowCellsLimit;
                    if (this.detectContainerUnlock()) {
                        this.isContainerLocked = false;
                    }
                }
            }
            else {
                this.slideLength = 0;
            }
        };
        Carousel.prototype.handleRightSlide = function () {
            if (this.slideCounter - this.slideLength < 0) {
                this.slideLength = this.slideCounter;
            }
            this.preliminarySlideCounter = this.slideCounter - this.slideLength;
            if (!this.isFirstCell) {
                this.newContainerPositionIndex = this.newContainerPositionIndex + this.slideLength;
                this.isSlideInProgress = true;
                if (this.isLazyLoad) {
                    if (this.preliminarySlideCounter > this.overflowCellsLimit) {
                        this.isContainerLocked = true;
                    }
                    else {
                        this.isContainerLocked = false;
                    }
                    if (this.detectContainerUnlock()) {
                        this.isContainerLocked = false;
                    }
                }
            }
            else {
                this.slideLength = 0;
            }
        };
        Carousel.prototype.getSlideLength = function () {
            var correction = this.getContainerPositionCorrection();
            var length = Math.floor((this.distanceAbs + correction) / this.fullCellWidth);
            if (this.distanceAbs % this.fullCellWidth >= this.minSwipeDistance) {
                length++;
            }
            return length;
        };
        /*
         * Limits the length of the slide during calls to the next() and prev()
         * methods if the specified position is outside the cell length
         */
        Carousel.prototype.limitSlideLength = function (slideLength) {
            if (slideLength > 1) {
                for (var i = 0; i < slideLength; i++) {
                    var newSlideCounter = this.slideCounter + (slideLength - i);
                    if (!this.detectLastSlide(newSlideCounter)) {
                        slideLength = slideLength - i;
                        this.isSlideLengthLimited = i > 0;
                        break;
                    }
                }
            }
            return slideLength;
        };
        Carousel.prototype.detectLastSlide = function (slideCounter) {
            if (this.properties.loop) {
                return false;
            }
            else {
                return (this.cellLength - slideCounter) < this.visibleCellsCount;
            }
        };
        Carousel.prototype.isNextArrowDisabled = function () {
            if (this.visibleCellsOverflowContainer) {
                return this.detectLastSlide(this.slideCounter + 1) && this.isContainerPositionCorrection;
            }
            else {
                return this.detectLastSlide(this.slideCounter + 1);
            }
        };
        Carousel.prototype.isPrevArrowDisabled = function () {
            return this.slideCounter === 0;
        };
        Carousel.prototype.detectContainerUnlock = function () {
            return (this.cellLength - this.preliminarySlideCounter) < (this.visibleCellsCount + this.overflowCellsLimit);
        };
        Carousel.prototype.handleSlideEnd = function () {
            if (this.isSlideInProgress) {
                this.transformSlideEnd();
                this.isSlideInProgress = false;
                this.newContainerPositionIndex = 0;
                this.isSlideLengthLimited = undefined;
            }
        };
        Carousel.prototype.transformSlideEnd = function () {
            if (this.isLazyLoad) {
                this.setSlideCounter();
                if (this.images || !this.images && this.properties.loop) {
                    this.quicklyPositionContainer();
                }
            }
            this.previousSlideCounter = this.slideCounter;
        };
        Carousel.prototype.setSlideCounter = function () {
            if (this.direction === 'left') {
                this.slideCounter = this.slideCounter + this.slideLength;
            }
            if (this.direction === 'right') {
                this.slideCounter = this.slideCounter - this.slideLength;
            }
            this.direction = undefined;
            this.slideLength = 0;
        };
        Carousel.prototype.resetTransition = function () {
            this.element.style.transition = '';
        };
        Carousel.prototype.getElementPosition = function () {
            return this.element.getBoundingClientRect();
        };
        Carousel.prototype.getCarouselElementPosition = function () {
            return this.properties.container.getBoundingClientRect();
        };
        Carousel.prototype.getDistance = function () {
            return this.startX - this.moveX;
        };
        Carousel.prototype.getDistanceAbs = function () {
            return Math.abs(this.startX - this.moveX);
        };
        Carousel.prototype.getDirection = function () {
            var direction = Math.sign(this.startX - this.moveX);
            if (direction === -1) {
                return 'right';
            }
            if (direction === 1) {
                return 'left';
            }
        };
        Carousel.prototype.lineUpCells = function () {
            var cells = this.element.children;
            for (var i = 0; i < cells.length; i++) {
                var cell = cells[i];
                var positionX = this.getCellPositionX(i);
                cell.style.transform = 'translateX(' + positionX + 'px)';
                // notice
                cell.style.width = this.properties.cellWidth + 'px';
            }
            ;
        };
        Carousel.prototype.getCellPositionX = function (index) {
            var positionIndex = this.getPositionIndex(index);
            return positionIndex * this.fullCellWidth;
        };
        Carousel.prototype.getPositionIndex = function (cellIndex) {
            var counter = this.slideCounter;
            var cellLength = this.totalContainerCellsCount;
            var slideCounter = counter - this.overflowCellsLimit;
            var positionIndex;
            if (slideCounter > cellLength) {
                slideCounter = slideCounter % cellLength;
            }
            if (slideCounter < 0) {
                return cellIndex;
            }
            else {
                positionIndex = cellIndex - slideCounter;
                if (positionIndex < 0) {
                    positionIndex = cellLength + positionIndex;
                }
            }
            return positionIndex;
        };
        Carousel.prototype.getCenterPositionIndex = function () {
            return (this.totalContainerCellsCount - 1) / 2;
        };
        Carousel.prototype.autoplay = function () {
            var _this = this;
            this.autoplayId = setInterval(function () {
                _this.next();
            }, this.properties.autoplayInterval);
        };
        Carousel.prototype.stopAutoplay = function () {
            if (this.autoplayId) {
                clearInterval(this.autoplayId);
            }
        };
        Carousel.prototype.destroy = function () {
            this.stopAutoplay();
        };
        return Carousel;
    }());

    var CarouselComponent = /** @class */ (function () {
        function CarouselComponent(elementRef, ref) {
            var _this = this;
            this.elementRef = elementRef;
            this.ref = ref;
            this.minTimeout = 30;
            this._cellWidth = 200;
            this.events = new core.EventEmitter();
            this.height = 200;
            this.loop = false;
            this.autoplay = false;
            this.autoplayInterval = 5000;
            this.pauseOnHover = true;
            this.dots = false;
            this.margin = 10;
            this.objectFit = 'cover';
            this.minSwipeDistance = 10;
            this.transitionDuration = 200;
            this.transitionTimingFunction = 'ease';
            this.counterSeparator = " / ";
            this.overflowCellsLimit = 3;
            this.listeners = 'mouse and touch';
            this.cellsToScroll = 1;
            this.arrows = true;
            this.arrowsTheme = 'light';
            this.hostClassCarousel = true;
            /* Touchstart */
            this.handleTouchstart = function (event) {
                //event.preventDefault();
                _this.touches.addEventListeners("mousemove", "handleMousemove");
                _this.carousel.handleTouchstart(event);
                _this.isMoving = true;
            };
            /* Touchmove */
            this.handleHorizontalSwipe = function (event) {
                event.preventDefault();
                _this.carousel.handleHorizontalSwipe(event);
            };
            /* Touchend */
            this.handleTouchend = function (event) {
                var touches = event.touches;
                _this.carousel.handleTouchend(event);
                _this.touches.removeEventListeners("mousemove", "handleMousemove");
                _this.isMoving = false;
            };
        }
        Object.defineProperty(CarouselComponent.prototype, "isContainerLocked", {
            get: function () {
                return this.carousel.isContainerLocked;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "slideCounter", {
            get: function () {
                return this.carousel.slideCounter;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "previousSlideCounter", {
            get: function () {
                return this.carousel.previousSlideCounter;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "lapCounter", {
            get: function () {
                return this.carousel.lapCounter;
            },
            enumerable: false,
            configurable: true
        });
        CarouselComponent.prototype.lineUpCells = function () {
            this.carousel.lineUpCells();
        };
        CarouselComponent.prototype.quicklyPositionContainer = function () {
            this.carousel.quicklyPositionContainer();
        };
        Object.defineProperty(CarouselComponent.prototype, "isLandscape", {
            get: function () {
                return window.innerWidth > window.innerHeight;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "isSafari", {
            get: function () {
                var ua = navigator.userAgent.toLowerCase();
                if (ua.indexOf('safari') !== -1) {
                    return !(ua.indexOf('chrome') > -1);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "counter", {
            get: function () {
                var counter;
                if (this.loop) {
                    counter = this.slideCounter % this.cellLength;
                }
                else {
                    counter = this.slideCounter;
                }
                return counter + 1 + this.counterSeparator + this.cellLength;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "cellsElement", {
            get: function () {
                return this.elementRef.nativeElement.querySelector('.carousel-cells');
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "images", {
            get: function () {
                return this._images;
            },
            set: function (images) {
                this._images = images;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "cellWidth", {
            set: function (value) {
                if (value) {
                    this._cellWidth = value;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "isCounter", {
            get: function () {
                return this._isCounter && this.cellLength > 1;
            },
            set: function (value) {
                if (value) {
                    this._isCounter = value;
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "activeDotIndex", {
            get: function () {
                return this.slideCounter % this.cellLength;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(CarouselComponent.prototype, "cellLimit", {
            get: function () {
                if (this.carousel) {
                    return this.carousel.cellLimit;
                }
            },
            enumerable: false,
            configurable: true
        });
        CarouselComponent.prototype.onWindowResize = function (event) {
            this.landscapeMode = this.isLandscape;
            this.ref.detectChanges();
            this.initCarousel();
            this.carousel.lineUpCells();
        };
        CarouselComponent.prototype.onMousemove = function (event) {
            if (this.autoplay && this.pauseOnHover) {
                this.carousel.stopAutoplay();
            }
        };
        CarouselComponent.prototype.onMouseleave = function (event) {
            if (this.autoplay && this.pauseOnHover) {
                this.carousel.autoplay();
            }
        };
        CarouselComponent.prototype.ngOnInit = function () {
            this.isNgContent = this.cellsElement.children.length > 0;
            this.touches = new Touches({
                element: this.cellsElement,
                listeners: this.listeners,
                mouseListeners: {
                    "mousedown": "handleMousedown",
                    "mouseup": "handleMouseup"
                }
            });
            this.touches.on('touchstart', this.handleTouchstart);
            this.touches.on('horizontal-swipe', this.handleHorizontalSwipe);
            this.touches.on('touchend', this.handleTouchend);
            this.touches.on('mousedown', this.handleTouchstart);
            this.touches.on('mouseup', this.handleTouchend);
            this.initCarousel();
            this.setDimensions();
            if (this.autoplay) {
                this.carousel.autoplay();
            }
        };
        CarouselComponent.prototype.ngAfterViewInit = function () {
            this.cellLength = this.getCellLength();
            this.dotsArr = Array(this.cellLength).fill(1);
            this.ref.detectChanges();
            this.carousel.lineUpCells();
            /* Start detecting changes in the DOM tree */
            this.detectDomChanges();
        };
        CarouselComponent.prototype.ngOnChanges = function (changes) {
            if (changes.width || changes.height || changes.images) {
                this.setDimensions();
                this.initCarousel();
                this.carousel.lineUpCells();
                this.ref.detectChanges();
            }
        };
        CarouselComponent.prototype.ngOnDestroy = function () {
            this.touches.destroy();
            this.carousel.destroy();
        };
        CarouselComponent.prototype.initCarousel = function () {
            this.carousel = new Carousel({
                element: this.elementRef.nativeElement.querySelector('.carousel-cells'),
                container: this.elementRef.nativeElement,
                images: this.images,
                cellWidth: this.getCellWidth(),
                loop: this.loop,
                autoplayInterval: this.autoplayInterval,
                overflowCellsLimit: this.overflowCellsLimit,
                visibleWidth: this.width,
                margin: this.margin,
                minSwipeDistance: this.minSwipeDistance,
                transitionDuration: this.transitionDuration,
                transitionTimingFunction: this.transitionTimingFunction,
                videoProperties: this.videoProperties
            });
        };
        CarouselComponent.prototype.detectDomChanges = function () {
            var _this = this;
            var observer = new MutationObserver(function (mutations) {
                _this.onDomChanges();
            });
            var config = {
                attributes: true,
                childList: true,
                characterData: true
            };
            observer.observe(this.elementRef.nativeElement, config);
        };
        CarouselComponent.prototype.onDomChanges = function () {
            this.cellLength = this.getCellLength();
            this.carousel.lineUpCells();
            this.ref.detectChanges();
        };
        CarouselComponent.prototype.setDimensions = function () {
            this.hostStyleHeight = this.height + 'px';
            this.hostStyleWidth = this.width + 'px';
        };
        CarouselComponent.prototype.getFile = function (index) {
            return this.carousel.getFile(index);
        };
        CarouselComponent.prototype.handleTransitionendCellContainer = function (event) {
            this.carousel.handleSlideEnd();
        };
        CarouselComponent.prototype.toggleVideo = function (video) {
            event.preventDefault();
            if (this.videoProperties.noPlay) {
                return;
            }
            if (video.paused) {
                video.play();
                this.isVideoPlaying = true;
            }
            else {
                video.pause();
                this.isVideoPlaying = false;
            }
            this.ref.detectChanges();
        };
        CarouselComponent.prototype.getCurrentIndex = function () {
            return this.carousel.slideCounter;
        };
        CarouselComponent.prototype.getCellWidth = function () {
            var elementWidth = this.elementRef.nativeElement.clientWidth;
            if (this.cellsToShow) {
                var margin = this.cellsToShow > 1 ? this.margin : 0;
                var totalMargin = margin * (this.cellsToShow - 1);
                return (elementWidth - totalMargin) / this.cellsToShow;
            }
            if (this._cellWidth === '100%') {
                return elementWidth;
            }
            else {
                return this._cellWidth;
            }
        };
        CarouselComponent.prototype.next = function () {
            this.carousel.next(this.cellsToScroll);
            this.carousel.stopAutoplay();
        };
        CarouselComponent.prototype.prev = function () {
            this.carousel.prev(this.cellsToScroll);
            this.carousel.stopAutoplay();
        };
        CarouselComponent.prototype.select = function (index) {
            this.carousel.select(index);
        };
        CarouselComponent.prototype.isNextArrowDisabled = function () {
            return this.carousel.isNextArrowDisabled();
        };
        CarouselComponent.prototype.isPrevArrowDisabled = function () {
            return this.carousel.isPrevArrowDisabled();
        };
        CarouselComponent.prototype.getCellLength = function () {
            if (this.images) {
                return this.images.length;
            }
            else {
                return this.cellsElement.children.length;
            }
        };
        return CarouselComponent;
    }());
    CarouselComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'carousel, [carousel]',
                    exportAs: 'carousel',
                    template: "<div class=\"carousel-counter\" *ngIf=\"isCounter\">{{counter}}</div>\n\n<div class=\"carousel-container\" [class.carousel-moving]=\"isMoving\">\n\t<div class=\"carousel-cells\" #cells (transitionend)=\"handleTransitionendCellContainer($event)\">\n\t\t<ng-content></ng-content>\n\n\t\t<ng-template ngFor let-image [ngForOf]=\"images\" let-i=\"index\">\n\t\t\t<div class=\"carousel-cell\" \n\t\t\t\t*ngIf=\"i < cellLimit\" \n\t\t\t\t[style.width]=\"getCellWidth()+'px'\"\n\t\t\t\t[style.border-radius]=\"borderRadius+'px'\">\n\n\t\t\t\t<!-- Image -->\n\t\t\t\t<img \n\t\t\t\t\t*ngIf=\"getFile(i) && getFile(i)['image']\" \n\t\t\t\t\t[src]=\"getFile(i)['image']['path']\"\n\t\t\t\t\t[style.object-fit]=\"objectFit\"\n\t\t\t\t\tdraggable=\"false\" />\n\n\t\t\t\t<!-- Video -->\n\t\t\t\t<ng-template [ngIf]=\"getFile(i) && getFile(i)['type'] === 'video'\">\n\t\t\t\t\t<div class=\"carousel-play\" (click)=\"toggleVideo(video)\"></div>\n\n\t\t\t\t\t<video #video\n\t\t\t\t\t\t[loop]=\"videoProperties.loop\"\n\t\t\t\t\t\t[muted]=\"videoProperties.muted\"\n\t\t\t\t\t\t[controls]=\"isSafari\"\n\t\t\t\t\t\t[style.object-fit]=\"objectFit\"\n\t\t\t\t\t\t[style.border-radius]=\"borderRadius+'px'\">\n\t\t\t\t\t\t\n\t\t\t\t\t\t<source [src]=\"getFile(i)['path']\" type=\"video/mp4\">\n\t\t\t\t\t</video>\n\t\t\t\t</ng-template>\n\t\t\t</div>\n\t\t</ng-template>\n\t</div>\n\n\t<div class=\"carousel-dots\" *ngIf=\"dots\">\n\t\t<div class=\"carousel-dot\" [class.carousel-dot-active]=\"i === activeDotIndex\" *ngFor=\"let dot of dotsArr; index as i\"></div>\n\t</div>\n</div>\n\n<div class=\"carousel-arrows\" \n\t[class.carousel-arrows-outside]=\"arrowsOutside\" \n\t[class.carousel-dark-arrows]=\"arrowsTheme === 'dark'\"\n\t*ngIf=\"arrows\">\n\t\n\t<div class=\"carousel-arrow carousel-arrow-prev\" [class.carousel-arrow-disabled]=\"isPrevArrowDisabled()\" (click)=\"prev()\"></div>\n\t<div class=\"carousel-arrow carousel-arrow-next\" [class.carousel-arrow-disabled]=\"isNextArrowDisabled()\" (click)=\"next()\"></div>\n</div>",
                    styles: [":host{-moz-user-select:none;-webkit-user-select:none;box-sizing:border-box;display:block;height:100%;left:0;position:relative;top:0;transform-origin:top left;user-select:none;width:100%;z-index:10000}:host .carousel-container{cursor:grab;height:100%;overflow:hidden;width:100%}:host .carousel-container.carousel-moving{cursor:grabbing}:host .carousel-counter{background-color:rgba(23,37,68,.3);border-radius:13px;color:#fff;font-size:11px;line-height:normal;padding:5px 7px;position:absolute;right:24px;text-align:right;top:8px;transition:opacity .2s;z-index:30}:host ::ng-deep .carousel-cells{display:block;height:100%;transition:transform .2s;width:100%}:host ::ng-deep .carousel-cells .carousel-cell.swiper-prev-image{transform:translate3d(-100%,0,0)}:host ::ng-deep .carousel-cells .carousel-cell.swiper-next-image{transform:translate3d(100%,0,0)}:host ::ng-deep .carousel-cells .carousel-cell{height:100%;overflow:hidden;position:absolute;width:100%}:host ::ng-deep .carousel-cells .carousel-cell img,:host ::ng-deep .carousel-cells .carousel-cell video{height:100%;object-fit:contain;position:relative;width:100%}:host ::ng-deep .carousel-cells .carousel-cell img.swiper-hide{display:none}:host ::ng-deep .carousel-cells .carousel-cell .carousel-play{bottom:0;left:0;position:absolute;right:0;top:0;z-index:1}:host .carousel-arrow{background-color:#fff;background-position:50%;background-repeat:no-repeat;background-size:31px;border-radius:100px;box-shadow:0 0 5px rgba(0,0,0,.15);cursor:pointer;height:40px;margin-top:-20px;position:absolute;top:50%;width:40px;z-index:10}:host .carousel-arrow-prev{background-image:url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgd2lkdGg9IjI0Ij48cGF0aCBkPSJNMTUuNDEgMTYuNTlMMTAuODMgMTJsNC41OC00LjU5TDE0IDZsLTYgNiA2IDYgMS40MS0xLjQxeiIvPjxwYXRoIGQ9Ik0wIDBoMjR2MjRIMFYweiIgZmlsbD0ibm9uZSIvPjwvc3ZnPg==);left:10px}:host .carousel-arrow-next{background-image:url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGhlaWdodD0iMjQiIHZpZXdCb3g9IjAgMCAyNCAyNCIgd2lkdGg9IjI0Ij48cGF0aCBkPSJNOC41OSAxNi41OUwxMy4xNyAxMiA4LjU5IDcuNDEgMTAgNmw2IDYtNiA2LTEuNDEtMS40MXoiLz48cGF0aCBkPSJNMCAwaDI0djI0SDBWMHoiIGZpbGw9Im5vbmUiLz48L3N2Zz4=);right:10px}:host .carousel-arrows-outside .carousel-arrow-prev{left:-60px}:host .carousel-arrows-outside .carousel-arrow-next{right:-60px}:host .carousel-dark-arrows .carousel-arrow{filter:invert(1)}:host .carousel-arrow-disabled{cursor:default;opacity:.5}:host .carousel-dots{bottom:0;left:0;position:absolute;right:0;text-align:center;z-index:10}:host .carousel-dots .carousel-dot{border:2px solid #fff;border-radius:100px;display:inline-block;height:8px;margin:4px;width:8px}:host .carousel-dots .carousel-dot-active{background-color:#fff}"]
                },] }
    ];
    CarouselComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };
    CarouselComponent.propDecorators = {
        images: [{ type: core.Input }],
        events: [{ type: core.Output }],
        height: [{ type: core.Input }],
        width: [{ type: core.Input }],
        loop: [{ type: core.Input }],
        autoplay: [{ type: core.Input }],
        autoplayInterval: [{ type: core.Input }],
        pauseOnHover: [{ type: core.Input }],
        dots: [{ type: core.Input }],
        borderRadius: [{ type: core.Input }],
        margin: [{ type: core.Input }],
        objectFit: [{ type: core.Input }],
        minSwipeDistance: [{ type: core.Input }],
        transitionDuration: [{ type: core.Input }],
        transitionTimingFunction: [{ type: core.Input }],
        videoProperties: [{ type: core.Input }],
        counterSeparator: [{ type: core.Input }],
        overflowCellsLimit: [{ type: core.Input }],
        listeners: [{ type: core.Input }],
        cellsToShow: [{ type: core.Input }],
        cellsToScroll: [{ type: core.Input }],
        cellWidth: [{ type: core.Input, args: ['cellWidth',] }],
        isCounter: [{ type: core.Input, args: ['counter',] }],
        arrows: [{ type: core.Input }],
        arrowsOutside: [{ type: core.Input }],
        arrowsTheme: [{ type: core.Input }],
        hostClassCarousel: [{ type: core.HostBinding, args: ['class.carousel',] }],
        hostStyleHeight: [{ type: core.HostBinding, args: ['style.height',] }],
        hostStyleWidth: [{ type: core.HostBinding, args: ['style.width',] }],
        onWindowResize: [{ type: core.HostListener, args: ['window:resize', ['$event'],] }],
        onMousemove: [{ type: core.HostListener, args: ['mousemove', ['$event'],] }],
        onMouseleave: [{ type: core.HostListener, args: ['mouseleave', ['$event'],] }]
    };

    var IvyCarouselModule = /** @class */ (function () {
        function IvyCarouselModule() {
        }
        return IvyCarouselModule;
    }());
    IvyCarouselModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        CarouselComponent
                    ],
                    imports: [
                        common.CommonModule
                    ],
                    exports: [
                        CarouselComponent
                    ],
                    providers: [],
                    bootstrap: [],
                    entryComponents: [
                        CarouselComponent
                    ]
                },] }
    ];

    /*
     * Public API Surface of angular-responsive-carousel
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CarouselComponent = CarouselComponent;
    exports.IvyCarouselModule = IvyCarouselModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
